### ISSUE TITLE
[Bug]: `react/prop-types` false positive when component is a value of capitalized property in class

### ISSUE THREAD

**Issue by Maxim-Mazurok** [yannickcr/eslint-plugin-react#3534]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

Weird error:
![image](https://user-images.githubusercontent.com/7756211/220268129-9387d92e-edf2-4249-adea-ece0d166bb9d.png)

```
import React from 'react';

class SomeComponent {
	column = {
		Cell: ({ original }: { original: string }) => <span>{original === 'bla' ? 'bla' : 'bla'}</span>,
	};
}
```

```
error    'original' is missing in props validation  react/prop-types
```

`eslint myfile.tsx`

### Expected Behavior

Not to throw error

When I change `Cell` to `cell` - the error is gone.
When I avoid using destructuring - the error is also gone.

### eslint-plugin-react version

v7.32.2

### eslint version

v8.34.0

### node version

v18.14.1

**Comment by ljharb** [comment]
`Cell` here is a function, taking props, that renders JSX, and doesn't have propTypes. Why is it named with a component naming convention if it's meant to be a render function?

**Comment by Maxim-Mazurok** [comment]
The reason it's named `Cell` is because it later passed to `react-table@6` as part of the [`columns` prop](https://github.com/TanStack/table/tree/v6.11.4#columns).

Here's a more complete example of how we use it:
```tsx
import React, { Component } from 'react';
import ReactTable from 'react-table';

export default class MyTable extends Component {
	columns = [
		{
			Cell: ({ original }: { original: string }) => <span>{original === 'bla' ? 'bla' : 'bla'}</span>,
		},
	];
	render(): JSX.Element {
		return <ReactTable columns={this.columns} />;
	}
}
```

Note that we're using `v6` of `react-table` which has a very different API from the latest version and no bandwidth to migrate.

**Comment by ljharb** [comment]
and ReactTable requires it to be named "Cell"? is that because it uses it like a component?

**Comment by Maxim-Mazurok** [comment]
> and ReactTable requires it to be named "Cell"?

Yes, see:
> Cell: JSX | String | Function // Used to render a standard cell, defaults to the accessed value

from [docs](https://github.com/TanStack/table/tree/v6.11.4#columns)
![image](https://user-images.githubusercontent.com/7756211/221051095-134a4853-d6d0-4862-b7e7-812c309f4736.png)


> is that because it uses it like a component?

I think so...

**Comment by ljharb** [comment]
right - so, it's a component. and components should have propTypes.

Is there a reason you can't do:
```jsx
import React from 'react';

class SomeComponent {
	column = {
		Cell({ original }: { original: string }) { return <span>{original === 'bla' ? 'bla' : 'bla'}</span>; },
	};
}
```
?

Components shouldn't be arrow functions, and I *think* that should be recognized by the prop-types rule as having prop types.

**Comment by Maxim-Mazurok** [comment]
> Is there a reason you can't do:
> 
> ```js
> import React from 'react';
> 
> class SomeComponent {
> 	column = {
> 		Cell({ original }: { original: string }) { return <span>{original === 'bla' ? 'bla' : 'bla'}</span>; },
> 	};
> }
> ```
> 
> ?

Your example doesn't give errors, but if I add do `class SomeComponent extends Component` - then it produces the same error:
![image](https://user-images.githubusercontent.com/7756211/221442411-a885bce2-1b0c-4206-9026-7e552f3de168.png)
`error  'original' is missing in props validation  react/prop-types`

> Components shouldn't be arrow functions

Surely this isn't a technical limitation (because our code works just fine), it's a stylistic preference, right?

**Comment by Maxim-Mazurok** [comment]
Hi, @ljharb just curious if you had time to look into this or have any thoughts.

**Comment by ljharb** [comment]
I haven't had time to look into it.

Yes, it's a stylistic preference, but it's more than just that, because an arrow function can't ever have an explicit name, and name inference isn't always intuitive.

**Comment by JMartinCollins** [comment]
I think I've experienced a similar issue. In my situation I have an HOC:
```tsx
const ModifiedComponent = withModifiedBehavior(Component, options);
```
The `options` object has the following type signature:
```tsx
interface WithModalOptions {
    // ...
    ModifiedBehaviorComponent?: React.ComponentType<{propWhichWillBePassed: () => void}>;
    // ...
}
```
`options` can take a property whose value is a React Component which will be rendered as part of the HOC modified behavior.
```tsx
const ModifiedComponent = withModifiedBehavior(Component, {
    ModifiedBehaviorComponent ({ propWhichWillBePassed }) {
        return (<>
            <div>Custom Text</div>
            <button 
                onClick={
                    // custom logic
                    propWhichWillBePassed()
                } 
            />
        </>)
    }
});
```
In this case `propWhichWillBePassed` has a squiggly line saying `'propWhichWillBePassed' is missing in props validationeslint[react/prop-types](https://github.com/jsx-eslint/eslint-plugin-react/tree/master/docs/rules/prop-types.md)`

The TS validation works perfectly fine here and it's able to infer the type and give all the necessary intellisense. I've tried declaring the `ModifiedBehaviorComponent` in the `options` object as a named function, like above, and also as arrow fn like this: `ModifiedBehaviorComponent: ({ propWhichWillBePassed  }) => { /* ... */ }`

In this case I'm not sure if there's a method for adding prop types to a Function Component provided as an object property.
